### ISSUE TITLE
Buggy results when compiled with Intel oneAPI DPC++/C++ Compiler

### ISSUE THREAD

**Issue by raphaeltrevizani** [IEDB/TCRMatch#43]
Changing the compiler to icpx (Intel's compiler, previously called icc) gave me different results, most with a score _greater than_ 1 which shouldn't happen). The results obtained with clang++ and g++ look fine and match each other.

Steps to reproduce:

Compile with Intel's compiler
icpx -fopenmp -O3 src/tcrmatch.cpp -o tcrmatch-icpx

Run it:
./tcrmatch-icpx -i test.txt -t1

To compare with the expected result:
Compile with g++
g++ -fopenmp -O3 src/tcrmatch.cpp -o tcrmatch-g++

Run it:
./tcrmatch-g++ -i test.txt -t1

The outputs I get are attached for each case. The input file (test.txt) contains a peptide from the database so it's score is 1 in the output files. Interestingly, the result obtained by icpx gives exactly 1 for this peptide, but it gives scores with values greater than 1 to others. 

[output.g++.txt](https://github.com/IEDB/TCRMatch/files/11064969/output.g%2B%2B.txt)
[output.icpx.txt](https://github.com/IEDB/TCRMatch/files/11064970/output.icpx.txt)
[test.txt](https://github.com/IEDB/TCRMatch/files/11064972/test.txt)

**Comment by bpeters42** [comment]
WOW! That is crazy. Definitely needs to be investigated, and it does not seem like an easy thing to follow up on. One suggestion I would have is to google for other people reporting differences in the compilers. And to make sure that the processors we are running it on are indeed the intel processors you have. Not sure how helpful this is. Please also reach out to others (including Morten) that have more actual experience than me with different compilers in this century.

**Comment by raphaeltrevizani** [comment]
Thanks for the input, @acrinklaw! Removing the optimization was one of the first things I tested, but without any success. I double checked it now and noticed that compiling with icc also gives different results when -O3 is used vs when -O3 is not used. When g++ is used, the results are the same for both. I suspect some uninitialized variable or I/O operations from unnalocated areas, as g++ usually deals with those pretty well.

**Comment by raphaeltrevizani** [comment]
I located where the bug is, but I'm not able to fix it because there doesn't seem to be anything wrong with the code. Henryk's comment [in this link](https://stackoverflow.com/questions/30263303/stdarray-vs-array-performance) matches what I observed.
I'll report more tomorrow after running more tests, but I'm convinced there's nothing wrong with TCRmatch. This seems to be a problem with Intel's compiler and how it handles std::array.

**Comment by danielmarrama** [comment]
May I suggest we add unit tests to the TCRMatch repo to capture stuff like this? You can run multiple compilers and check the expected output within GitHub.

**Comment by acrinklaw** [comment]
> May I suggest we add unit tests to the TCRMatch repo to capture stuff like this? You can run multiple compilers and check the expected output within GitHub.

yeah I had meant to do this 2 years back but you know how things go...

I just made a branch "feature/unit_tests" that incorporates [catch2](https://github.com/catchorg/Catch2) and a GitHub action to get some basic CI/CD going.

I made a dummy unit test that should fail just as an example
![image](https://user-images.githubusercontent.com/25108033/229040163-161b9446-aede-47f3-a74e-f17bd7ddfa14.png)

I'd be more than happy to keep working on this in my free time. Just let me know if you want to coordinate @raphaeltrevizani I don't want to overstep any bounds or make the code harder for you to work on

**Comment by raphaeltrevizani** [comment]
Replacing the `std::array<std:array<float,20>, 20>` by basic c-arrays takes care of the problem and the code compiled with icpx now matches g++. It is indeed a problem with Intel's compiler as we suspected. I uploaded the code with c-arrays to the c_array branch and I'm closing this issue. Thanks for the help, all!
